### PR TITLE
o/snapstate: add support for kernel-modules handlers

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
Add handlers and backend methods that will set-up early mount units and kernel tree when installing a kernel-modules component.

This is mostly the same as https://github.com/snapcore/snapd/pull/13616, but now using directly the mounts in "/snap" instead of creating new mounts in "/run/mnt" for kernel and kernel components, as agreed recently. Therefore mount units for these snaps/components have new Before= dependencies.

### CODE REVIEW

File: systemd/systemd.go [hunk 1]
```diff
@@ -1432,6 +1437,7 @@ What={{.What}}
 Where={{.Where}}
 Type={{.Fstype}}
 Options={{join .Options ","}}
+LazyUnmount=yes
```
Comment (valentindavid) [inline systemd/systemd.go:1440]: I see all the other generate mount units have it, so it is fine. But I think one day we should remove all the `LazyUnmount` in all the generated units. I do not think they are needed. And if we have bugs, we should fix them another way.

File: systemd/systemd.go [hunk 2]
```diff
@@ -1416,14 +1419,16 @@ X-SnapdOrigin={{.}}
...
+// We want some snaps and components to be mounted before modules are
 // loaded (that is before systemd-{udevd,modules-load}).
 const beforeDriversLoadUnitTmpl = `[Unit]
 Description={{.Description}}
 DefaultDependencies=no
 After=systemd-remount-fs.service
 Before=sysinit.target
 Before=systemd-udevd.service systemd-modules-load.service
+Before=snapd.mounts.target
+Before=local-fs.target
```
Comment (valentindavid) [inline systemd/systemd.go:1431]: I am trying to figure out now, why we cannot use the same template, and only add `Before=systemd-udevd.service systemd-modules-load.service` only for kernel module snaps.  I do not see the dependency that needs to be removed.

File: overlord/snapstate/backend/setup.go [hunk 3]
```diff
@@ -348,10 +297,10 @@ func (b Backend) SetupKernelModulesComponents(compsToInstall, currentComps []*sn
 	sysd := newSystemd(b.preseed, meter)
 
 	// compsToRemove will contain the old revisions of components in compsToInstall
```
Comment (pedronis) [inline overlord/snapstate/backend/setup.go:299]: comment needs changing

File: overlord/snapstate/backend/setup.go [hunk 4]
```diff
@@ -397,109 +346,27 @@ func mergeCompSideInfosUpdatingRev(comps1, comps2 []*snap.ComponentSideInfo) (me
 
 // moveKModsComponentsState changes kernel-modules set-up from currentComps to
 // finalComps, and adds/removes toAdd and toRemove components respectively.
```
Comment (pedronis) [inline overlord/snapstate/backend/setup.go:348]: same here

File: overlord/snapstate/snapmgr.go [hunk 5]
```diff
@@ -1302,9 +1303,14 @@ func (m *SnapManager) ensureMountsUpdated() error {
...
 			// an update.
+			// TODO refactor so the check for kernel type is not repeated
+			// in the installation case
+			snapType, _ := snapSt.Type()
 			if _, err = sysd.EnsureMountUnitFile(info.MountDescription(),
 				squashfsPath, whereDir, "squashfs",
-				systemd.EnsureMountUnitFlags{PreventRestartIfModified: true}); err != nil {
+				systemd.EnsureMountUnitFlags{
+					PreventRestartIfModified: true,
+					StartBeforeDriversLoad:   snapType == snap.TypeKernel}); err != nil {
```
Comment (pedronis) [inline overlord/snapstate/snapmgr.go:1313]: not quite related to the PR but do we need to add code here about components mounts as well?

**Review by pedronis — COMMENTED** [review]
thanks, a few comments on the changes from the previous PR

File: overlord/snapstate/backend/setup.go (hunk 3 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/backend/setup.go]: Done

File: overlord/snapstate/backend/setup.go (hunk 4 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/backend/setup.go]: Done

File: overlord/snapstate/snapmgr.go (hunk 5 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/snapmgr.go:1314]: Right, I've added a TODO about that

File: systemd/systemd.go (hunk 2 above)
Comment (alfonsosanchezbeato) [reply systemd/systemd.go:1431]: There is an `After=snapd.mounts-pre.target` in the regular mount units that probably we don't want here. Anyway, even if that was not the case and we could have very similar units by using `DefaultDependencies=no` and adding manually the dependencies I prefer to have separate templates precisely so regular units do not need to be explicit about dependencies and they look simpler.

As a side note, I've read again systemd.mount(5) and I have changed:
```
After=systemd-remount-fs.service
```
to
```
After=local-fs-pre.target
```
as that seems to be the default dependency (should not make a big difference).

File: systemd/systemd.go (hunk 1 above)
Comment (alfonsosanchezbeato) [reply systemd/systemd.go:1440]: Right, I agree

**Review by alfonsosanchezbeato — COMMENTED** [review]
Thanks for the reviews, I've addressed the comments now

File: overlord/snapstate/handlers_components.go [hunk 6]
```diff
@@ -402,3 +402,72 @@ func (m *SnapManager) undoUnlinkCurrentComponent(t *state.Task, _ *tomb.Tomb) (e
...
+	st.Lock()
+	compSetup, snapsup, snapSt, err := compSetupAndState(t)
+	st.Unlock()
+	if err != nil {
+		return err
+	}
+
+	// kernel-modules components already in the system
+	st.Lock()
+	kmodComps := snapSt.Sequence.ComponentsWithTypeForRev(snapsup.Revision(), snap.KernelModulesComponent)
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply overlord/snapstate/handlers_components.go:420]: Good point, this is indeed a deep copy, so the lock can be removed

File: overlord/snapstate/backend/setup.go [hunk 7]
```diff
@@ -325,3 +289,85 @@ func (b Backend) UndoSetupComponent(cpi snap.ContainerPlaceInfo, installRecord *
...
+			// No change for this component
+			merged = append(merged, instComp)
+		}
+	}
+
+	return merged, instersect
+}
+
+// moveKModsComponentsState changes kernel-modules set-up from currentComps to
+// finalComps, for the kernek/revision specified by ksnapName/ksnapRev.
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply overlord/snapstate/backend/setup.go]: Done

File: overlord/snapstate/backend/mountunit_test.go [hunk 8]
```diff
@@ -96,6 +97,11 @@ func (s *mountunitSuite) TearDownTest(c *C) {
 }
 
 func (s *mountunitSuite) TestAddMountUnit(c *C) {
+	s.testAddMountUnit(c, systemd.EnsureMountUnitFlags{})
+	s.testAddMountUnit(c, systemd.EnsureMountUnitFlags{StartBeforeDriversLoad: true})
+}
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply overlord/snapstate/backend/mountunit_test.go]: Done

File: overlord/snapstate/backend/setup.go [hunk 9]
```diff
@@ -325,3 +289,85 @@ func (b Backend) UndoSetupComponent(cpi snap.ContainerPlaceInfo, installRecord *
...
+		currentActiveComps, finalComps, ksnapName, ksnapRev, sysd,
+		"after failure to remove set-up of kernel modules components")
+}
+
+// mergeCompSideInfosUpdatingRev returns a merged list from two lists of
+// ComponentSideInfo, and a list with the intersection, using the criteria of
+// the elements having the same ComponentRef. The rest of the data for an
+// element will come from comps2 if ComponentRef is the same in comps1 and
+// comps2, that is, the revision is updated in that case.
+func mergeCompSideInfosUpdatingRev(comps1, comps2 []*snap.ComponentSideInfo) (merged, instersect []*snap.ComponentSideInfo) {
```
Comment (bboozzoo) [inline overlord/snapstate/backend/setup.go:325]: btw. interesect isn't used by any of the callers, is there a PR in queue which will make use of it? if not, I'd drop it for now.

**Review by bboozzoo — APPROVED** [review]
LGTM

File: systemd/systemd.go (hunk 2 above)
Comment (valentindavid) [reply systemd/systemd.go:1431]: My point is that I do not think that `DefaultDependencies=no` is needed. I do not see what introduced dependencies are breaking it.

Could we know why `After=snapd.mounts-pre.target` is bad for this one? I do not think it has to be after module loading.

Comment (alfonsosanchezbeato) [reply systemd/systemd.go]: Ok, I got you know. Yeah, definitely in the end I am repeating the default dependencies in the early mount units. I have changed things so we use only one template. Re: `After=snapd.mounts-pre.target` I think we can have it also for early mounts. It has an after dependency on `zfs-mount.service` due to https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/1922293 , which should be fine.

File: overlord/snapstate/backend/setup.go (hunk 9 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/backend/setup.go]: Yeah, it is not used now, I've removed in the end.

File: systemd/systemd.go [hunk 10]
```diff
@@ -1416,33 +1422,13 @@ X-SnapdOrigin={{.}}
...
-Type={{.Fstype}}
-Options={{join .Options ","}}
-
-[Install]
-WantedBy=sysinit.target
-{{- with .Origin}}
-X-SnapdOrigin={{.}}
-{{- end}}
-`
+func IsBeforeDriversLoadMountUnit(mType MountUnitType) bool {
```
Comment (pedronis) [inline systemd/systemd.go:1425]: do we need to export this? if we do it needs a comment

**Review by pedronis — APPROVED** [review]
thanks, small comment/question

File: systemd/systemd.go (hunk 10 above)
Comment (alfonsosanchezbeato) [reply systemd/systemd.go]: Not really, for some reason I thought that maybe the  go template would need it exported, but is not the case. Unexported now.

File: systemd/systemd.go [hunk 11]
```diff
@@ -1398,11 +1398,14 @@ var squashfsFsType = squashfs.FsType
 
 // Note that WantedBy=multi-user.target and Before=local-fs.target are
 // only used to allow downgrading to an older version of snapd.
-const regularMountUnitTmpl = `[Unit]
+//
+// We want (see isbeforedrivers) some snaps and components to be mounted before
```
Comment (bboozzoo) [inline systemd/systemd.go:1402]:
```suggestion
// We want (see isBeforeDrivers) some snaps and components to be mounted before
```
or simply drop the function name. Function names in comments have a tendency to become outdated and incorrect whenever refactors are introduced.

**Review by bboozzoo — APPROVED** [review]
LGTM

File: systemd/systemd.go (hunk 11 above)
Comment (alfonsosanchezbeato) [reply systemd/systemd.go]: Done, thanks

**Review by valentindavid — APPROVED** [review]
Thank you.

**Comment by alfonsosanchezbeato** [comment]
Thanks for the reviews, I've squashed some commits now

**Pull request merged**